### PR TITLE
Optimize jpeg decoder based on traces.

### PR DESCRIPTION
**Pull request by JimBobSquarePants**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description

I ran a trace against the decoder after reading this benchmark [here](https://github.com/kleisauke/net-vips/tree/master/tests/NetVips.Benchmarks) (_which we're doing really well with IMO_) and thought I'd clean up some low hanging fruit.  Shaved a few percent of the decoding benchmarks.

**Decode**

Adds hardware accelerated rounding to `SimdUtils.FastRound`.

Before and after performance traces. 
jd-before
jd-after

|                         Method |            TestImage |       Mean |       Error |     StdDev | Ratio | RatioSD | Gen 0 | Gen 1 | Gen 2 |  Allocated |
|------------------------------- |--------------------- |-----------:|------------:|-----------:|------:|--------:|------:|------:|------:|-----------:|
| 'Decode Jpeg - System.Drawing' | Jpg/b(...)e.jpg [21] |   5.122 ms |   1.3978 ms |  0.0766 ms |  1.00 |    0.00 |     - |     - |     - |      176 B |
|     'Decode Jpeg - ImageSharp' | Jpg/b(...)e.jpg [21] |  11.991 ms |   0.2514 ms |  0.0138 ms |  2.34 |    0.03 |     - |     - |     - |    15816 B |
|                                |                      |            |             |            |       |         |       |       |       |            |
| 'Decode Jpeg - System.Drawing' | Jpg/b(...)f.jpg [28] |  14.943 ms |   1.8410 ms |  0.1009 ms |  1.00 |    0.00 |     - |     - |     - |      176 B |
|     'Decode Jpeg - ImageSharp' | Jpg/b(...)f.jpg [28] |  29.759 ms |   1.5452 ms |  0.0847 ms |  1.99 |    0.01 |     - |     - |     - |    16824 B |
|                                |                      |            |             |            |       |         |       |       |       |            |
| 'Decode Jpeg - System.Drawing' | Jpg/i(...)e.jpg [43] | 388.229 ms | 382.8946 ms | 20.9877 ms |  1.00 |    0.00 |     - |     - |     - |      216 B |
|     'Decode Jpeg - ImageSharp' | Jpg/i(...)e.jpg [43] | 276.490 ms | 195.5104 ms | 10.7166 ms |  0.71 |    0.01 |     - |     - |     - | 36022368 B |

**Scan Only**
Slight refactor of `HuffmanScanBuffer.DecodeHuffman` to remove double lookup on slow path. There's now nothing that stands out here. 
image

|                      Method |        Job |       Runtime |            TestImage |     Mean |     Error |    StdDev | Ratio |   Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------------------------- |----------- |-------------- |--------------------- |---------:|----------:|----------:|------:|--------:|------:|------:|----------:|
|       'System.Drawing FULL' | Job-HITJFX |    .NET 4.7.2 | Jpg/b(...)e.jpg [21] | 5.828 ms | 0.9885 ms | 0.0542 ms |  1.00 | 46.8750 |     - |     - |  211566 B |
| JpegDecoderCore.ParseStream | Job-HITJFX |    .NET 4.7.2 | Jpg/b(...)e.jpg [21] | 5.833 ms | 0.2923 ms | 0.0160 ms |  1.00 |       - |     - |     - |   12416 B |
|                             |            |               |                      |          |           |           |       |         |       |       |           |
|       'System.Drawing FULL' | Job-WPSKZD | .NET Core 2.1 | Jpg/b(...)e.jpg [21] | 6.018 ms | 2.1374 ms | 0.1172 ms |  1.00 | 46.8750 |     - |     - |  210768 B |
| JpegDecoderCore.ParseStream | Job-WPSKZD | .NET Core 2.1 | Jpg/b(...)e.jpg [21] | 4.382 ms | 0.9009 ms | 0.0494 ms |  0.73 |       - |     - |     - |   12360 B |
|                             |            |               |                      |          |           |           |       |         |       |       |           |
|       'System.Drawing FULL' | Job-ZLSNRP | .NET Core 3.1 | Jpg/b(...)e.jpg [21] | 5.714 ms | 0.4078 ms | 0.0224 ms |  1.00 |       - |     - |     - |     176 B |
| JpegDecoderCore.ParseStream | Job-ZLSNRP | .NET Core 3.1 | Jpg/b(...)e.jpg [21] | 4.239 ms | 1.0943 ms | 0.0600 ms |  0.74 |       - |     - |     - |   12406 B |

### CODE REVIEW

File: tests/ImageSharp.Benchmarks/Codecs/Jpeg/DecodeJpegParseStreamOnly.cs [hunk 1]
```diff
@@ -15,7 +14,7 @@ namespace SixLabors.ImageSharp.Benchmarks.Codecs.Jpeg
     [Config(typeof(Config.ShortClr))]
     public class DecodeJpegParseStreamOnly
     {
-        [Params(TestImages.Jpeg.Baseline.Jpeg420Exif)]
+        [Params(TestImages.Jpeg.BenchmarkSuite.Lake_Small444YCbCr)]
```
Comment (JimBobSquarePants) [inline tests/ImageSharp.Benchmarks/Codecs/Jpeg/DecodeJpegParseStreamOnly.cs:17]: Updated to match the full decoding benchmark so we can more easily see the difference.

File: src/ImageSharp/Common/Helpers/SimdUtils.cs [hunk 2]
```diff
@@ -44,13 +48,24 @@ internal static Vector4 PseudoRound(this Vector4 v)
...
-            Vector<float> sub0 = Vector.Subtract(add0, or0);
-            return sub0;
+#if SUPPORTS_RUNTIME_INTRINSICS
+
+            if (Avx.IsSupported)
+            {
+                ref Vector256<float> v256 = ref Unsafe.As<Vector<float>, Vector256<float>>(ref v);
+                Vector256<float> vRound = Avx.RoundToNearestInteger(v256);
+                return Unsafe.As<Vector256<float>, Vector<float>>(ref vRound);
+            }
```
Comment (antonfirsov) [inline src/ImageSharp/Common/Helpers/SimdUtils.cs:58]:
```suggestion
            if (Avx2.IsSupported)
            {
                ref Vector256<float> v256 = ref Unsafe.As<Vector<float>, Vector256<float>>(ref v);
                Vector256<float> vRound = Avx.RoundToNearestInteger(v256);
                return Unsafe.As<Vector256<float>, Vector<float>>(ref vRound);
            }
```

There is a call into this method from `Vector4Pair.RoundAndDownscalePreVector8` feeding 4-element 128 bit vectors the method, therefore We need to make sure that on old CPU-s having `Vector<float>.Count == 4`, this path is never executed.

If I remember it correctly AVX2 is needed for `Vector<float>.Count == 8`, so with that check we would be safe. @saucecontrol can you confirm?

Comment (saucecontrol) [reply src/ImageSharp/Common/Helpers/SimdUtils.cs:58]: > AVX2 is needed for Vector<float>.Count == 8

Yup :)

Comment (JimBobSquarePants) [reply src/ImageSharp/Common/Helpers/SimdUtils.cs:58]: I was actually wondering about that. Thanks!

**Pull request merged**